collect_app/src/main/java/org/odk/collect/android/activities/FormEntryActivity.java

```diff
@@ -2891,5 +2882,81 @@ public void onReceive(Context context, Intent intent) {
         }
     }
 
+    @Override
+    public void widgetValueChanged(QuestionWidget changedWidget) {
+        FormController formController = Collect.getInstance().getFormController();
+        if (formController == null) {
+            // TODO: As usual, no idea if/how this is possible.
+            return;
+        }
+
+        if (formController.indexIsInFieldList()) {
+            updateFieldListQuestions(changedWidget.getFormEntryPrompt().getIndex());
+        }
+    }
+
+    /**
+     * Saves the form and updates displayed widgets accordingly:
+     * - removes widgets corresponding to questions that are no longer relevant
+     * - adds widgets corresponding to questions that are newly-relevant
+     * - removes and rebuilds widgets corresponding to questions that have changed in some way. For
+     * example, the question text or hint may have updated due to a value they refer to changing.
+     *
+     * The widget corresponding to the {@param lastChangedIndex} is never changed.
+     */
+    private void updateFieldListQuestions(FormIndex lastChangedIndex) {
```

Comment:
The logic here comparing beforeSave vs. afterSave makes sense, and there's no urgent need to change it, but my two cents:

The way React handles updates like this -- [described fabulously in this detailed blog post](https://overreacted.io/react-as-a-ui-runtime/#lists) -- is basically by having a key for each element and comparing keys in order to determine what actually changed (vs. what simply changed order but doesn't need to be re-rendered). If we give each question a UUID, then map them below using this UUID instead of their formIndex, that could help to more easily determine additions/subtractions from the UI.

It might also just throw a wrench in an already-working system, so no hard feelings.

Reply:
That's a really good blog post, thank you for sharing.

I'm not immediately seeing a good way to build those keys and have them be meaningful across form saves. That is, different objects get mutated and that results in computed values like label text being different. That's why I do that gymnastics with `ImmutableDisplayableQuestion` and `sameAs`. Anything I can come up with would require doing something getting all of what I save in `ImmutableDisplayableQuestion` to compute the key. Did you see a good option off the top of your head?

Reply:
The benefit I'm trying to get from this is in the case where a new question is added above several other questions -- in that case it seems like the FormIndex of the old questions would change (can you confirm if this is true?)

questionsBeforeSave = [a, b, c]
questionsAfterSave = [a, x, b, c]

`b` and `c` don't need to re-render, but their FormIndex may have changed. They are technically `sameAs` each other but `questionAtSameFormIndex` is different so they aren't ever compared. If they were assigned some ephemeral UUID (doesn't need to be persisted anywhere), they could be matched using that UUID instead of the FormIndex.

Is this even a problem worth solving? Now that I write it down I realize this is an exceptional case -- the "typical" case to optimize for is when you're typing in a text field and **no** re-renders are triggered. So this change wouldn't help in the typical case.

Reply:
`FormIndex` is based on the position in the *form definition* and so isn't affected by relevance and the like so I think we're doing basically what you suggest with the `FormIndex` being the UUID.